File: Host/rtime.java
import java.io.IOException;
import java.util.GregorianCalendar;

public class rtime {
	private static enum Format {
		DECIMAL,
		HEX,
		BINARY,
		COMMAND;
	}
	
	private static String h8(long v) {
		String h = "00000000" + Long.toHexString(v).toUpperCase();
		return h.substring(h.length() - 8);
	}
	
	public static void main(String[] args) throws IOException {
		Format format = Format.DECIMAL;
		int argi = 0;
		while (argi < args.length) {
			String arg = args[argi++];
			if (arg.startsWith("-") && arg.length() >= 2) {
				switch (arg.charAt(1)) {
				case 'd':
					format = Format.DECIMAL;
					break;
				case 'h':
					format = Format.HEX;
					break;
				case 'b':
					format = Format.BINARY;
					break;
				case 'r':
					format = Format.COMMAND;
					break;
				default:
					System.out.println("Usage: java rtime [-d|-h|-b|-r]");
					return;
				}
			} else {
				System.out.println("Usage: java rtime [-d|-h|-b|-r]");
				return;
			}
		}
		
		GregorianCalendar c = new GregorianCalendar();
		long days = c.getTimeInMillis() / 86400000;
		long msec = c.getTimeInMillis() % 86400000;
		long tzn = c.get(GregorianCalendar.ZONE_OFFSET);
		long dst = c.get(GregorianCalendar.DST_OFFSET);
		
		switch (format) {
		case DECIMAL:
			System.out.println(days);
			System.out.println(msec);
			System.out.println(tzn);
			System.out.println(dst);
			break;
		case HEX:
			System.out.println(h8(days));
			System.out.println(h8(msec));
			System.out.println(h8(tzn));
			System.out.println(h8(dst));
			break;
		case BINARY:
			System.out.write(new byte[] {
				(byte)(days >> 24L), (byte)(days >> 16L), (byte)(days >> 8L), (byte)(days),
				(byte)(msec >> 24L), (byte)(msec >> 16L), (byte)(msec >> 8L), (byte)(msec),
				(byte)(tzn >> 24L), (byte)(tzn >> 16L), (byte)(tzn >> 8L), (byte)(tzn),
				(byte)(dst >> 24L), (byte)(dst >> 16L), (byte)(dst >> 8L), (byte)(dst),
			});
			System.out.flush();
			break;
		case COMMAND:
			System.out.write(new byte[] {
				'r', 0x11, 0, 0,
				(byte)(days >> 24L), (byte)(days >> 16L), (byte)(days >> 8L), (byte)(days),
				'r', 0x13, 0, 0,
				(byte)(msec >> 24L), (byte)(msec >> 16L), (byte)(msec >> 8L), (byte)(msec),
				'r', 0x14, 0, 0,
				(byte)(tzn >> 24L), (byte)(tzn >> 16L), (byte)(tzn >> 8L), (byte)(tzn),
				'r', 0x15, 0, 0,
				(byte)(dst >> 24L), (byte)(dst >> 16L), (byte)(dst >> 8L), (byte)(dst),
			});
			System.out.flush();
			break;
		}
	}
}

File: trunk/Host/rtime.c
#include <stdio.h>
#include "rclock.h"

#define DECIMAL 0
#define HEX     1
#define BINARY  2
#define COMMAND 3
#define FIELDS  4

void write_int(int v) {
	char buf[4];
	buf[0] = (char)(v >> 24);
	buf[1] = (char)(v >> 16);
	buf[2] = (char)(v >> 8);
	buf[3] = (char)(v);
	fwrite(buf, 1, 4, stdout);
}

int main(int argc, char ** argv) {
	long int days;
	long int msec;
	long int tzn;
	long int dst;
	
	int format = DECIMAL;
	int argi = 1;
	while (argi < argc) {
		char * arg = argv[argi++];
		if (arg[0] == '-' && arg[1]) {
			switch (arg[1]) {
			case 'd':
				format = DECIMAL;
				break;
			case 'h':
				format = HEX;
				break;
			case 'b':
				format = BINARY;
				break;
			case 'r':
				format = COMMAND;
				break;
			case 'f':
				format = FIELDS;
				break;
			default:
				printf("Usage: rtime [-d|-h|-b|-r|-f]\n");
				return 1;
			}
		} else {
			printf("Usage: rtime [-d|-h|-b|-r|-f]\n");
			return 1;
		}
	}
	
	days = get_clock_hi(1);
	msec = get_clock_lo(0);
	tzn = get_clock_tzn(0);
	dst = get_clock_dst(0);
	
	switch(format) {
	case DECIMAL:
		printf("%ld\n", days);
		printf("%ld\n", msec);
		printf("%ld\n", tzn);
		printf("%ld\n", dst);
		break;
	case HEX:
		printf("%08X\n", (int)days);
		printf("%08X\n", (int)msec);
		printf("%08X\n", (int)tzn);
		printf("%08X\n", (int)dst);
		break;
	case BINARY:
		write_int((int)days);
		write_int((int)msec);
		write_int((int)tzn);
		write_int((int)dst);
		fflush(stdout);
		break;
	case COMMAND:
		write_int(0x72110000);
		write_int((int)days);
		write_int(0x72130000);
		write_int((int)msec);
		write_int(0x72140000);
		write_int((int)tzn);
		write_int(0x72150000);
		write_int((int)dst);
		fflush(stdout);
		break;
	case FIELDS:
		for (argi = 0; argi < 64; argi++) {
			printf("%2d: %d\n", argi, (int)get_clock(0, argi));
		}
		break;
	}
	
	return 0;
}
